pinot-core/src/main/java/com/linkedin/pinot/core/realtime/converter/stats/RealtimeColumnStatistics.java

```diff
@@ -0,0 +1,156 @@
+
+import com.linkedin.pinot.common.data.FieldSpec;
+import com.linkedin.pinot.core.common.Block;
+import com.linkedin.pinot.core.common.BlockMultiValIterator;
+import com.linkedin.pinot.core.data.partition.PartitionFunction;
+import com.linkedin.pinot.core.io.reader.SingleColumnSingleValueReader;
+import com.linkedin.pinot.core.operator.blocks.RealtimeSingleValueBlock;
+import com.linkedin.pinot.core.realtime.impl.datasource.RealtimeColumnDataSource;
+import com.linkedin.pinot.core.realtime.impl.dictionary.MutableDictionaryReader;
+import com.linkedin.pinot.core.segment.creator.ColumnStatistics;
+import java.util.List;
+import org.apache.commons.lang.math.IntRange;
+
+
+/**
+ * Column statistics for a column coming from an in-memory realtime segment.
+ */
+public class RealtimeColumnStatistics implements ColumnStatistics {
+  private final RealtimeColumnDataSource _dataSource;
+  private final int[] _sortedDocIdIterationOrder;
+  private final MutableDictionaryReader _dictionaryReader;
+  private final Block _block;
+
+  public RealtimeColumnStatistics(RealtimeColumnDataSource dataSource, int[] sortedDocIdIterationOrder) {
+    _dataSource = dataSource;
+    _sortedDocIdIterationOrder = sortedDocIdIterationOrder;
+    _dictionaryReader = dataSource.getDictionary();
+    _block = dataSource.getNextBlock();
+  }
+
+  @Override
+  public Object getMinValue() {
+    return _dictionaryReader.getMinVal();
+  }
+
+  @Override
+  public Object getMaxValue() {
+    return _dictionaryReader.getMaxVal();
+  }
+
+  @Override
+  public Object getUniqueValuesSet() {
+    return _dictionaryReader.getSortedValues();
+  }
+
+  @Override
+  public int getCardinality() {
+    return _dictionaryReader.length();
+  }
+
+  @Override
+  public int getLengthOfLargestElement() {
+    // Length of longest string
+    int maximumStringLength = 0;
+
+    // If this column is a string column, iterate over the dictionary to find the maximum length
+    if (_dataSource.getDataSourceMetadata().getDataType() == FieldSpec.DataType.STRING) {
+      final int length = _dictionaryReader.length();
+      for (int i = 0; i < length; i++) {
+        maximumStringLength = Math.max(_dictionaryReader.getStringValue(i).length(), maximumStringLength);
+      }
+    }
+
+    return maximumStringLength;
+  }
+
+  @Override
+  public boolean isSorted() {
+    // Multivalue columns can't be in sorted order
+    if (!_block.getMetadata().isSingleValue()) {
+      return false;
+    }
+
+    // If this is a single value, then by definition the data is sorted
+    final int blockLength = _block.getMetadata().getLength();
+    if (blockLength <= 1 || getCardinality() <= 1) {
+      return true;
+    }
+
+    // Iterate over all data to figure out whether or not it's in sorted order
+    SingleColumnSingleValueReader singleValueReader = ((RealtimeSingleValueBlock) _block).getReader();
+
+    int docIdIndex = _sortedDocIdIterationOrder != null ? _sortedDocIdIterationOrder[0] : 0;
```

Comment:
If it's extracted into a method, then the optimizer can't notice that this check is actually a loop invariant without inlining that method.